ext/liquid_c/vm_assembler.c

```diff
@@ -73,6 +74,113 @@ void vm_assembler_gc_mark(vm_assembler_t *code)
+        switch (*ip) {
+            case OP_LEAVE:
+                rb_str_catf(output, "leave\n");
+                break;
+
+            case OP_POP_WRITE:
+                rb_str_catf(output, "pop_write\n");
+                break;
+
+            case OP_PUSH_NIL:
+                rb_str_catf(output, "push_nil\n");
+                break;
+
+            case OP_PUSH_TRUE:
+                rb_str_catf(output, "push_true\n");
+                break;
+
+            case OP_PUSH_FALSE:
+                rb_str_catf(output, "push_false\n");
+                break;
+
+            case OP_FIND_VAR:
+                rb_str_catf(output, "find_var\n");
+                break;
+
+            case OP_LOOKUP_KEY:
+                rb_str_catf(output, "lookup_key\n");
+                break;
+
+            case OP_NEW_INT_RANGE:
+                rb_str_catf(output, "new_int_range\n");
+                break;
+
+            case OP_HASH_NEW:
+                rb_str_catf(output, "hash_new(%u)\n", ip[1]);
+                break;
+
+            case OP_PUSH_INT8:
+                rb_str_catf(output, "push_int8(%u)\n", ip[1]);
+                break;
+
+            case OP_PUSH_INT16:
+            {
+                int num = *(int8_t *)ip++;
+                num = (num << 8) | ip[1];
+                rb_str_catf(output, "push_int16(%u)\n", num);
+                break;
+            }
+
+            case OP_RENDER_VARIABLE_RESCUE:
+            {
+                unsigned int line_number = decode_node_line_number(ip + 1);
+                rb_str_catf(output, "render_variable_rescue(line_number: %u)\n", line_number);
+                break;
+            }
+
+            case OP_WRITE_RAW:
+            {
+                const char *text = (const char *)const_ptr[0];
+                size_t size = const_ptr[1];
+                VALUE string = rb_enc_str_new(text, size, utf8_encoding);
+                rb_str_catf(output, "write_raw(%+"PRIsVALUE")\n", string);
+                break;
+            }
+
+            case OP_WRITE_NODE:
+                rb_str_catf(output, "write_node(%+"PRIsVALUE")\n", const_ptr[0]);
+                break;
+
+            case OP_PUSH_CONST:
+                rb_str_catf(output, "push_const(%+"PRIsVALUE")\n", const_ptr[0]);
+                break;
+
+            case OP_FIND_STATIC_VAR:
+                rb_str_catf(output, "find_static_var(%+"PRIsVALUE")\n", const_ptr[0]);
+                break;
+
+            case OP_LOOKUP_CONST_KEY:
+                rb_str_catf(output, "lookup_const_key(%+"PRIsVALUE")\n", const_ptr[0]);
+                break;
+
+            case OP_LOOKUP_COMMAND:
+                rb_str_catf(output, "lookup_command(%+"PRIsVALUE")\n", const_ptr[0]);
+                break;
+
+            case OP_FILTER:
+                rb_str_catf(output, "filter(name: %+"PRIsVALUE", num_args: %u)\n", const_ptr[0], ip[1]);
+                break;
+
+            default:
+                rb_str_catf(output, "<%d>\n", ip[0]);
```

Comment:
Should we output a more obvious error message here?

Reply:
It isn't exactly an error, since it is safe to show the opcode number if the opcode is handled in `liquid_vm_next_instruction`.  However, you're right that it wasn't clear why just an number is shown like that.  I pushed a commit to change it to `rb_str_catf(output, "<opcode number %d disassembly not implemented>\n", ip[0]);`